src/deploy/hosting/uploader.ts

```diff
@@ -0,0 +1,268 @@
+import { size } from "lodash";
+import AbortController from "abort-controller";
+import * as clc from "cli-color";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as path from "path";
+import * as zlib from "zlib";
+
+import { Client } from "../../apiv2";
+import { Queue } from "../../throttler/queue";
+import { hostingApiOrigin } from "../../api";
+import * as hashcache from "./hashcache";
+import * as logger from "../../logger";
+import { FirebaseError } from "../../error";
+
+const MIN_UPLOAD_TIMEOUT = 30000; // 30s
+const MAX_UPLOAD_TIMEOUT = 7200000; // 2h
+
+function progressMessage(message: string, current: number, total: number): string {
+  current = Math.min(current, total);
+  const percent = Math.floor(((current * 1.0) / total) * 100).toString();
+  return `${message} [${current}/${total}] (${clc.bold.green(`${percent}%`)})`;
+}
+
+export class Uploader {
+  private version: string;
+  private cwd: string;
+  private projectRoot: string;
+  private gzipLevel: number;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```

Comment:
nit: Can we consider `/* eslint-disable @typescript-eslint/no-explicit-any */` and then re-enable a few lines after? A disable comment every two lines seems really noisy.

Reply:
ack. I've found using disable blocks far easier to set and then forget to unset, and I'd rather have the pain of excluding any `any` individually because they _really_ should have a specific type (but in converting js->ts, I'm not usually bothered to fix all of them)